ydb/core/kafka_proxy/actors/kafka_balancer_actor.cpp

```diff
@@ -0,0 +1,774 @@
+    KAFKA_LOG_D("HandleHeartbeat request");
+
+    auto heartbeatRequest = ev->Get()->Request;
+
+    RequestType   = HEARTBEAT;
+    CurrentStep   = STEP_NONE;
+    CorellationId = ev->Get()->CorrelationId;
+
+    GroupId  = heartbeatRequest->GroupId.value();
+    MemberId = heartbeatRequest->MemberId.value();
+
+    Kqp->SendCreateSessionRequest(ctx);
+}
+
+
+void TKafkaBalancerActor::Handle(NKqp::TEvKqp::TEvCreateSessionResponse::TPtr& ev, const TActorContext& ctx) {
+    const TString createSessionError = "Failed to create KQP session";
+    if (!Kqp->HandleCreateSessionResponse(ev, ctx)) {
+        switch (RequestType) {
+            case JOIN_GROUP:
+                SendJoinGroupResponseFail(ctx, CorellationId,
+                                          EKafkaErrors::UNKNOWN_SERVER_ERROR,
+                                          createSessionError);
+                break;
+            case SYNC_GROUP:
+                SendSyncGroupResponseFail(ctx, CorellationId,
+                                          EKafkaErrors::UNKNOWN_SERVER_ERROR,
+                                          createSessionError);
+                break;
+            case LEAVE_GROUP:
+                SendLeaveGroupResponseFail(ctx, CorellationId,
+                                           EKafkaErrors::UNKNOWN_SERVER_ERROR,
+                                           createSessionError);
+                break;
+            case HEARTBEAT:
+                SendHeartbeatResponseFail(ctx, CorellationId,
+                                          EKafkaErrors::UNKNOWN_SERVER_ERROR,
+                                          createSessionError);
+                break;
+            default:
+                break;
+        }
+        PassAway();
+        return;
+    }
+
+    switch (RequestType) {
+        case JOIN_GROUP:
+            HandleJoinGroupResponse(nullptr, ctx);
+            break;
+        case SYNC_GROUP:
+            HandleSyncGroupResponse(nullptr, ctx);
+            break;
+        case LEAVE_GROUP:
+            HandleLeaveGroupResponse(nullptr, ctx);
+            break;
+        case HEARTBEAT:
+            HandleHeartbeatResponse(nullptr, ctx);
+            break;
+        default:
+            KAFKA_LOG_CRIT("Unknown RequestType in TEvCreateSessionResponse");
+            PassAway();
+            break;
+    }
+}
+
+void TKafkaBalancerActor::Handle(NKqp::TEvKqp::TEvQueryResponse::TPtr& ev, const TActorContext& ctx) {
+    const TString kqpQueryError = "KQP query error";
+    if (ev->Cookie != KqpReqCookie) {
+        KAFKA_LOG_CRIT("Unexpected cookie in TEvQueryResponse");
+        return;
+    }
+
+    const auto& record = ev->Get()->Record;
+    auto status   = record.GetYdbStatus();
+    auto kafkaErr = KqpStatusToKafkaError(status);
+
+    if (kafkaErr != EKafkaErrors::NONE_ERROR) {
+        switch (RequestType) {
+            case JOIN_GROUP:
+                SendJoinGroupResponseFail(ctx, CorellationId, kafkaErr, kqpQueryError);
```

Comment:
Refactor this code. Make SendResponseFail() function and switch with RequestType inside. Otherwise there is copy-paste.